File: frontend/lib/models/bill.dart
// lib/models/bill.dart
class Bill {
  final String id;
  final String user;
  final String mess;
  final int month;
  final int year;
  final double baseAmount;
  final double rebateAmount;
  final double totalAmount;
  final String status; // 'Due' | 'Pending Approval' | 'Paid'
  final String? paymentProofUrl;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  Bill({
    required this.id,
    required this.user,
    required this.mess,
    required this.month,
    required this.year,
    required this.baseAmount,
    required this.rebateAmount,
    required this.totalAmount,
    required this.status,
    this.paymentProofUrl,
    this.createdAt,
    this.updatedAt,
  });

  factory Bill.fromJson(Map<String, dynamic> json) {
    return Bill(
      id: json['_id'] as String,
      user: json['user'] as String,
      mess: json['mess'] as String,
      month: json['month'] as int,
      year: json['year'] as int,
      baseAmount: (json['baseAmount'] as num).toDouble(),
      rebateAmount: (json['rebateAmount'] as num).toDouble(),
      totalAmount: (json['totalAmount'] as num).toDouble(),
      status: json['status'] as String,
      paymentProofUrl: json['paymentProofUrl'] as String?,
      createdAt:
          json['createdAt'] != null ? DateTime.parse(json['createdAt']) : null,
      updatedAt:
          json['updatedAt'] != null ? DateTime.parse(json['updatedAt']) : null,
    );
  }

  Map<String, dynamic> toJson() => {
        '_id': id,
        'user': user,
        'mess': mess,
        'month': month,
        'year': year,
        'baseAmount': baseAmount,
        'rebateAmount': rebateAmount,
        'totalAmount': totalAmount,
        'status': status,
        if (paymentProofUrl != null) 'paymentProofUrl': paymentProofUrl,
        if (createdAt != null) 'createdAt': createdAt!.toIso8601String(),
        if (updatedAt != null) 'updatedAt': updatedAt!.toIso8601String(),
      };
}

File: backend/models/Bill.js
const mongoose = require('mongoose');

const billSchema = new mongoose.Schema({
  user: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  mess: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Mess',
    required: true
  },
  month: {
    type: Number,
    required: true,
    min: 1,
    max: 12
  },
  year: {
    type: Number,
    required: true
  },
  baseAmount: {
    type: Number,
    required: true,
    min: 0
  },
  rebateAmount: {
    type: Number,
    default: 0,
    min: 0
  },
  totalAmount: {
    type: Number,
    required: true,
    min: 0
  },
  status: {
    type: String,
    enum: ['Due', 'Pending Approval', 'Paid'],
    default: 'Due'
  },
  paymentProofUrl: {
    type: String
  }
}, {
  timestamps: true
});

// Compound indexes for efficient queries
billSchema.index({ user: 1, mess: 1, month: 1, year: 1 }, { unique: true });
billSchema.index({ mess: 1, status: 1 });
billSchema.index({ mess: 1, status: 1, year: -1, month: -1, updatedAt: -1 });
billSchema.index({ user: 1, mess: 1, year: -1, month: -1, createdAt: -1 });

module.exports = mongoose.model('Bill', billSchema);
